### DISCUSSION TITLE
proposal: Semantic Release

### DISCUSSION THREAD

**Discussion by Xuanwo** [databendlabs/databend#6383] [Ideas]
## Background

`fusebots` used to help us make releases every day. However, we don't want to maintain an extra service after the adoption of mergify. Instead, we want to generate changelogs and upload releases via GitHub actions instead.

Also, our community has a need for `semantic release` which allows us to generate more user-friendly releases which contain `refactor` or `bugfix`.

## Proposal

So I propose to change our release process in the following ways:

### Enable semantic PRs

We will adopt semantic PR via `mergify`: all PR's titles should carry a tag. PRs that do not carry tags will be rejected by `mergify`

Within this change:

- We will remove the requirement of filling `Changelog` in the PR description: they will be inferred by PR's tag
- We will apply PR labels like `pr-bugfix` via PR's tag

For example:

- `bugfix: show query with limit will failed whend enable planner v2` will be labeled as `pr-bugfix` and recorded as `Bugfix` in changedlog
- `feat(ast): add span info for TableReference` will be labeled as `pr-feat` and recorded as `New Features` in the changelog
- `ci: Upgrade rust version` will be labeled as `pr-build` and ignored in changelog

### Enable semantic release

We will make release notes via GitHub native API: [Automatically generated release notes](https://docs.github.com/en/repositories/releasing-projects-on-github/automatically-generated-release-notes).

By maintaining `.github/release.yml`, we will generate changelogs by labels of PRs.

### Why not `semantic-release`?

[semantic-release](https://github.com/semantic-release/semantic-release) is a great project but:

- It's too complex
- Not designed for rust projects (we don't have a `package.json`)
- Commits based (we need changelogs for PR instead)

### Valid tags

- `feat`: this PR introduces a new feature to the codebase
- `fix`: this PR patches a bug in codebase
- `refactor`: this PR changes the code base without new features or bugfix
- `ci|build`: this PR changes build/testing/ci steps
- `docs|website`: this PR changes the documents or websites
- `chore`: this PR only has small changes that no need to record, like coding styles.

**Comment by PsiACE** [comment]
LGTM, Would we consider adding a copy of `CHANGELOG.md` into the repo?

**Reply by Xuanwo** [reply]
LGTM, we can do this in the future.

**Comment by drmingdrmer** [comment]
>`bugfix: show query with limit will failed whend enable planner v2` will be labeled as pr-bugfix and recorded as Bugfix in changedlog


Which part of the PR does it extract keywords from?
Commit message subject, commit message body, PR title, PR body?

**Reply by PsiACE** [reply]
I think this should support extraction from PR title, provided it meets some sort of specification.

**Reply by Xuanwo** [reply]
Only in pr title
